File: src/main/java/club/sk1er/mods/core/universal/UniversalResolutionUtil.java
package club.sk1er.mods.core.universal;

//#if FORGE && MC<11500
import net.minecraft.client.gui.ScaledResolution;
//#endif

public class UniversalResolutionUtil {
    private static final UniversalResolutionUtil INSTANCE = new UniversalResolutionUtil();

    //#if FORGE && MC<11500
    private ScaledResolution scaledResolution;
    private int cachedWidth = 0;
    private int cachedHeight = 0;
    //#endif

    private UniversalResolutionUtil() {
    }

    public static UniversalResolutionUtil getInstance() {
        return INSTANCE;
    }

    public int getWindowWidth() {
        //#if FABRIC
        //$$ return UniversalMinecraft.getMinecraft().getWindow().getWidth();
        //#elseif MC>=11502
        //$$ return UniversalMinecraft.getMinecraft().getMainWindow().getWidth();
        //#else
        return UniversalMinecraft.getMinecraft().displayWidth;
        //#endif
    }

    public int getWindowHeight() {
        //#if FABRIC
        //$$ return UniversalMinecraft.getMinecraft().getWindow().getHeight();
        //#elseif MC>=11502
        //$$ return UniversalMinecraft.getMinecraft().getMainWindow().getHeight();
        //#else
        return UniversalMinecraft.getMinecraft().displayHeight;
        //#endif
    }

    //#if MC<11502
    private ScaledResolution get() {
        if (cachedHeight != getWindowHeight() || cachedWidth != getWindowWidth() || scaledResolution == null) {
            scaledResolution = new ScaledResolution(UniversalMinecraft.getMinecraft());
        }
        return scaledResolution;
    }
    //#endif

    public int getScaledWidth() {
        //#if FABRIC
        //$$ return UniversalMinecraft.getMinecraft().getWindow().getScaledWidth();
        //#elseif MC>=11502
        //$$ return UniversalMinecraft.getMinecraft().getMainWindow().getScaledWidth();
        //#else
        return get().getScaledWidth();
        //#endif
    }

    public int getScaledHeight() {
        //#if FABRIC
        //$$ return UniversalMinecraft.getMinecraft().getWindow().getScaledHeight();
        //#elseif MC>=11502
        //$$ return UniversalMinecraft.getMinecraft().getMainWindow().getScaledHeight();
        //#else
        return get().getScaledHeight();
        //#endif
    }

    public int getScaleFactor() {
        //#if FABRIC
        //$$ return (int) UniversalMinecraft.getMinecraft().getWindow().getScaleFactor();
        //#elseif MC>=11502
        //$$ return (int)UniversalMinecraft.getMinecraft().getMainWindow().getGuiScaleFactor();
        //#else
        return get().getScaleFactor();
        //#endif
    }
}

File: src/main/kotlin/club/sk1er/mods/core/universal/UniversalResolutionUtil.kt
package club.sk1er.mods.core.universal

//#if MC<=11202
import net.minecraft.client.gui.ScaledResolution
//#endif

object UniversalResolutionUtil {
    //#if MC<=11202
    private var scaledResolution: ScaledResolution? = null
    private var cachedWidth: Int? = null
    private var cachedHeight: Int? = null
    //#endif

    @JvmStatic
    val windowWidth: Int
        get() {
            //#if FABRIC
            //$$ return UniversalMinecraft.getMinecraft().window.width
            //#elseif MC>=11502
            //$$ return UniversalMinecraft.getMinecraft().mainWindow.width
            //#else
            return UniversalMinecraft.getMinecraft().displayWidth
            //#endif
        }

    @JvmStatic
    val windowHeight: Int
        get() {
            //#if FABRIC
            //$$ return UniversalMinecraft.getMinecraft().window.height
            //#elseif MC>=11502
            //$$ return UniversalMinecraft.getMinecraft().mainWindow.height
            //#else
            return UniversalMinecraft.getMinecraft().displayHeight
            //#endif
        }

    //#if MC<=11202
    private fun get(): ScaledResolution {
        if (cachedHeight != windowHeight || cachedWidth != windowWidth || scaledResolution == null)
            scaledResolution = ScaledResolution(UniversalMinecraft.getMinecraft())
        return scaledResolution!!
    }
    //#endif

    @JvmStatic
    val scaledWidth: Int
        get() {
            //#if FABRIC
            //$$ return UniversalMinecraft.getMinecraft().window.scaledWidth
            //#elseif MC>=11502
            //$$ return UniversalMinecraft.getMinecraft().mainWindow.scaledWidth
            //#else
            return get().scaledWidth
            //#endif
        }

    @JvmStatic
    val scaledHeight: Int
        get() {
            //#if FABRIC
            //$$ return UniversalMinecraft.getMinecraft().window.scaledHeight
            //#elseif MC>=11502
            //$$ return UniversalMinecraft.getMinecraft().mainWindow.scaledHeight
            //#else
            return get().scaledHeight
            //#endif
        }

    @JvmStatic
    val scaleFactor: Double
        get() {
            //#if FABRIC
            //$$ return UniversalMinecraft.getMinecraft().window.scaleFactor
            //#elseif MC>=11502
            //$$ return UniversalMinecraft.getMinecraft().mainWindow.guiScaleFactor
            //#else
            return get().scaleFactor.toDouble()
            //#endif
        }
}
